KEYCLOAK-18146 Search for clients by client attribute when doing saml…

mhajas opened the pull request:
… artifact resolution

<!---
Please read https://github.com/keycloak/keycloak/blob/master/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

mhajas committed:
KEYCLOAK-18146 Search for clients by client attribute when doing saml artifact resolution

hmlnarik commented on model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate14_0_0_MigrateSamlArtifactAttribute.java:
```diff
@@ -0,0 +1,97 @@
+
+    private boolean hasSamlAttributeIdentifier(String idOfClient) throws CustomChangeException {
+        try (PreparedStatement statement = jdbcConnection.prepareStatement("SELECT NAME FROM " + getTableName("CLIENT_ATTRIBUTES") +
+                " WHERE CLIENT_ID=? AND NAME=?")) {
```
```suggestion
                " WHERE CLIENT_ID=? AND CLIENT_ID=?")) {
```

hmlnarik commented on model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate14_0_0_MigrateSamlArtifactAttribute.java:
```diff
@@ -0,0 +1,97 @@
+
+    @Override
+    protected void generateStatementsImpl() throws CustomChangeException {
+        extractClientsData("SELECT ID, CLIENT_ID FROM " + getTableName("CLIENT") + " " +
+                "WHERE PROTOCOL='saml'");
```
```suggestion
        extractClientsData("SELECT c.ID, c.CLIENT_ID FROM " + getTableName("CLIENT") + " c " +
                " LEFT JOIN " + getTableName("CLIENT_ATTRIBUTES") + " ca ON c.CLIENT_ID = ca.CLIENT_ID AND ca.name = '" + SAML_ARTIFACT_BINDING_IDENTIFIER + "' " +
                "WHERE c.PROTOCOL='saml' AND ca.name IS NULL");
```

mhajas replied:
Thanks, that is nice improvement. I will add it.

hmlnarik commented on model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate14_0_0_MigrateSamlArtifactAttribute.java:
```diff
@@ -0,0 +1,97 @@
+        }
+    }
+
+    private boolean hasSamlAttributeIdentifier(String idOfClient) throws CustomChangeException {
+        try (PreparedStatement statement = jdbcConnection.prepareStatement("SELECT NAME FROM " + getTableName("CLIENT_ATTRIBUTES") +
+                " WHERE CLIENT_ID=? AND NAME=?")) {
+            statement.setString(1, idOfClient);
+            statement.setString(2, SAML_ARTIFACT_BINDING_IDENTIFIER);
+            try (ResultSet rs = statement.executeQuery()) {
+                return rs.next(); // The client has saml attribute
+            }
+        } catch (Exception e) {
+            throw new CustomChangeException(getTaskId() + ": Exception when extracting data from previous version", e);
+        }
+    }
+
```
This method can be removed if the query above is used.

hmlnarik commented on server-spi-private/src/main/java/org/keycloak/models/utils/RepresentationToModel.java:
```diff
@@ -1406,6 +1408,11 @@ private static ClientModel createClient(KeycloakSession session, RealmModel real
             }
         }
 
+        if ("saml".equals(resourceRep.getProtocol())
```
Prefer `SamlProtocol.LOGIN_PROTOCOL` (other places as well)

mhajas replied:
`SamlProtocol` is part of services and `RepresentationToModel` is in `server-spi-private` so it is not available there. I can create some constants class in `server-spi-private` and then reuse it everywhere, but I am not sure this is what we want.

hmlnarik replied:
Ah, that makes perfect sense :)

hmlnarik commented on testsuite/integration-arquillian/tests/other/console/src/test/java/org/keycloak/testsuite/console/clients/AbstractClientTest.java:
```diff
@@ -87,6 +89,7 @@ public static ClientRepresentation createClientRep(String clientId, String proto
         attributes.put(SAML_SIGNATURE_ALGORITHM, "RSA_SHA256");
         attributes.put(SAML_FORCE_NAME_ID_FORMAT, "false");
         attributes.put(SAML_NAME_ID_FORMAT, "username");
+        attributes.put(SamlConfigAttributes.SAML_ARTIFACT_BINDING_IDENTIFIER, ArtifactBindingUtils.computeArtifactBindingIdentifierString("saml"));
```
Is this line really necessary? `SAML_ARTIFACT_BINDING_IDENTIFIER` should be automatically set on server when not present

mhajas replied:
If I understand the test correctly, these attributes are just checked after the creation, they are not set when creating the client.

hmlnarik replied:
It seems they are used also for creation, see e.g. `org.keycloak.testsuite.console.clients.ClientSettingsTest.createSAML()`
This line can stay here if there is another test (e.g. in base testsuite) that tests the creation of SAML client without this attribute and then checks that the attribute has been added by keycloak server when reading the client back.

mhajas replied:
From `newClient` `ClientRepresentation` only `clientId` and `protocol` is used for creation, see [this](https://github.com/keycloak/keycloak/blob/master/testsuite/integration-arquillian/tests/other/console/src/main/java/org/keycloak/testsuite/console/page/clients/CreateClientForm.java#L27-L31).

hmlnarik reviewed (changes requested):
Overall looks good to me @mhajas! I believe the only thing that needs fixing is the migration.

mhajas commented:
@hmlnarik @vramik I updated the PR based on your feedback. I have also run Jenkins again, I will post the results once it is finished. 

Here are previous results:
https://keycloak-jenkins.com/job/universal-test-pipeline-server/449/
https://keycloak-jenkins.com/job/universal-test-pipeline-adapters/396/

I fixed the failures of `JsonFileImport***MigrationTest` that was caused by my changes. I cannot identify anything else, but if you see anything else just let me know.

mhajas commented:
Last Jenkins runs:
https://master-jenkins.com/job/universal-test-pipeline-server/454
https://master-jenkins.com/job/universal-test-pipeline-adapters/403/

hmlnarik reviewed (approved):
Thank you for the PR!
